### ISSUE TITLE
RFE: JSON-API backup/restore - AOS 1.1/AOS 1.2

### ISSUE THREAD

**Issue by jpsenior** [Apstra/aos-pyez#91]
From time to time we may need to perform a 'breaking change' AOS controller upgrade.  This means wiping the AOS Database and starting over.  

Some resources are able to be carried from the old version to the new as new REST API posts to various APIs.  This will be helpful during POC processes and in the field when customers are incrementally upgrading their stuff.

This request is for a generic 'backup' and 'restore' functionality with the following criteria:

Enumerate IP pools and ASN pools, store as JSON backup.
Enumerate External routers, store as JSON backup
Enumerate HCL, store as JSON backup
Enumerate Logical devices, store as JSON backup
Enumerate Logical device maps, store as JSON backup
Enumerate Configlets, store as JSON backup
Enumerate Templates, store as JSON backup
Enumerate Rack types, store as JSON backup

The 'restore' function should connect to AOS and enumerate existing resources - if a conflict on id or display_name exists, don't re-post the json backup file.

This does not cover potential changes to AOS REST API and modelling semantics.  If a POST fails, would be nice to have a --continue function.

**Comment by jeremyschulman** [comment]
@jpsenior - agree this would be a nice utility.  I am putting together a "hit list" of such utilities.  I have one that easily allows someone to create new HCL entries, for example.

That said, the existing aos-pyez library supports the underpinning functions that you need.  Each collection item has both a `loads` and `dumps` method for exactly the purpose you describe.  The missing part, of course, is the CLI utility.

That said, you could fabricate an Ansible playbook to do what you want; at least for now if you happen to be blocked.  Alternatively, I would be very happy to accept a pull request if you want to create the CLI utility yourself.  And I could provide whatever guidance and review necessary.

**Comment by dgarros** [comment]
I think the real challenge will be to backup and restore the blueprints. 

For the resources listed below, it's very easy to do that with Ansible and the aos_ modules we have developed. I already have some playbook to restore most of the resources listed. It would be easy use Ansible as well to save all resources in Yaml or JSON.

**Comment by jpsenior** [comment]
One of the challenges I can see is trying to understand the semantics of differences between GET and POST - e.g. the GET includes resource pool allocation statistics (in use, not in use, etc) which are not valid with the same POST.

**Comment by jpsenior** [comment]
I should also mention that I cannot use an ansible playbook on a Windows host, which is a requirement for some of our customers

**Comment by dgarros** [comment]
Why that ? 
Can you use docker ?

**Comment by jpsenior** [comment]
Ansible doesn't run on Windows at all.  This is a core design philosophy on core ansible.  There are no reliable workarounds. We can run ansible 'against' windows, but not 'from windows.

**Comment by dgarros** [comment]
We can if we use docker :)

**Comment by jpsenior** [comment]
If you can pull something off, I'll be the first to cheer in happiness, it's something that I've been blocked on for years now....  Have to so far maintain random linux VMS to only simply run playbooks for quite a long time...

**Comment by dgarros** [comment]
It's already there, we even have [a docker container with ansible + all dependancies needed to use it with an AOS server.](https://hub.docker.com/r/apstra/ansible/) 
1. [Install docker for Windows on your laptop](https://docs.docker.com/docker-for-windows/install/)
2. run `docker run -it -v <local dir>:/scripts -w /scripts apstra/ansible sh` to get into the container in cli

`-v <local dir>:/scripts` will mount your local directory into the container and you can execute ansible inside the container.

Let me know if you need help

**Comment by jeremyschulman** [comment]
closing due to alternative approach taken.
